Add parsing for signed payload

noch3 opened the pull request:
Looking for initial thoughts.

jsdw commented on core_v14/src/decoder/mod.rs:
```diff
@@ -307,10 +313,45 @@ pub struct ExtrinsicSignature {
-	let address = <MultiAddress<AccountId32, u32>>::decode(data)?;
-	let signature = MultiSignature::decode(data)?;
-	let extensions = metadata
+fn decode_call_data<'data, 'meta>(
```
I've been hesitant to make too many things `pub` at the moment becaue this V14 crate is still fairly heavily WIP, so I've been learning towards not committing too much to the public API, and making sure that we don't expose too many things going forwards and make it a little harder to keep hacking on without breaking APIs.

w.r.t the parsing, I'd sortof have assumed that the top level parser + scale_info would be enough (given some type and some bytes, you'll get a decoded `Value` back), but I'd be interested to hear about any cases where the sub parsers were useful on their own!

noch3 replied:
In my experience, instead of being afraid to break APIs, you can just put anything that isn't really stable inside of a `unstable` module. That way people (like me) who are ok depending on unstable features can do so without patching your library. Anyone who cares about that can avoid the module. This, to me, is the best of both worlds.

noch3 replied:
The problem with the top-level parsers is that you can't compose them yourself. If I have call-data inside of a larger blob, it's harder to use your parser (although I can pull it out of the LateEof error in this case). But, for example, I had to manually write `decode_call_data` here because I couldn't compose that with the other parsers. If I had access to that directly I may not have patched this library (although you could argue it was a good thing that the patch got upstreamed).

jsdw commented:
Thank you! It might take me a few days to get to looking at this, but it looks like something worth wrapping my head around :)

noch3 commented:
> Thank you! It might take me a few days to get to looking at this, but it looks like something worth wrapping my head around :)

The SignedPayload (or SignerPayload, as it's called in polkadot-js) is just `[call-data][signed-extensions][extra-signed]`. You were already parsing `call-data` and `signed-extensions` so I just copied the code for `signed-extensions` and changed it to parse `extra-signed`. Then I refactored the code to make it easy to compose these various parsers together to get the full payload parser.

jsdw commented:
Ah, I understand! Basically, you'd like to expose the ability to decode the payload that would be signed; got it!

I had a chat with @insipx and I think we're arriving at a decision on an API for this, so if it's OK with you @noch3, I'll push a couple of commits to this PR tommorow (most likely) to shuffle a couple of bits around and keep the API consistent and such :)

noch3 commented:
If these commits are blocked by the CLA, you may not want to build on them directly.

insipx commented:
> If these commits are blocked by the CLA, you may not want to build on them directly.

Have you signed CLA or is it just not working?

noch3 commented:
> I'm working with my employer to sign the CLA. It might take a while.

noch3 commented:
If you want to reimplement this on your own I won't be even slightly offended. Just so long as I can use it. :D

jsdw commented:
I've opened #66 as my attempt to take this code, plus the request for more information form your previous PR, and mash them together into something :)

noch3 commented:
Closing in favor of #66
